File: competition-serv-service/src/main/java/compman/compsrv/service/schedule/StageGraph.kt
package compman.compsrv.service.schedule

import com.compmanager.compservice.jooq.tables.pojos.FightDescription
import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import compman.compsrv.model.dto.brackets.StageDescriptorDTO
import org.slf4j.LoggerFactory
import java.math.BigDecimal

class StageGraph(stages: List<StageDescriptorDTO>, fights: List<FightDescription>) {

    companion object {
        private val log = LoggerFactory.getLogger(StageGraph::class.java)
    }

    private val stagesGraph: Array<List<Int>>
    private val fightsGraph: Array<List<Int>>
    private val stageIdsToIds: BiMap<String, Int>
    private val fightIdsToIds: BiMap<String, Int>
    private val fightIdToCategoryId: MutableMap<String, String> = mutableMapOf()
    private val fightIdToStageId: MutableMap<String, String> = mutableMapOf()
    private val fightIdToDuration: MutableMap<String, BigDecimal> = mutableMapOf()
    private val stageIdToFightIds: Array<MutableSet<String>>
    private val stageOrdering: IntArray
    private val fightsOrdering: IntArray
    private var nonCompleteCount: Int
    private val fightsInDegree: IntArray
    private val completedFights: BooleanArray
    private val completableFights: MutableSet<Int>
    private val categoryIdToFightIds: Map<String, Set<String>>



    init {

        categoryIdToFightIds = fights.groupBy { it.categoryId }.mapValues { e -> e.value.map { it.id }.toSet() }
        stageIdsToIds = HashBiMap.create()
        fightIdsToIds = HashBiMap.create()
        // we resolve transitive connections here (a -> b -> c ~  (a -> b, a -> c, b -> c))
        var i = 0
        for (stage in stages) {
            if (!stageIdsToIds.containsKey(stage.id)) {
                stageIdsToIds[stage.id] = i++
            }
        }
        val stageNodesMutable = Array<MutableSet<Int>>(i) { HashSet() }
        stageIdToFightIds = Array<MutableSet<String>>(i) { HashSet() }

        stages.forEach { stage ->
            stage.inputDescriptor?.selectors?.forEach { s ->
                s.applyToStageId?.let { parentId ->
                    if (stageIdsToIds.containsKey(parentId)) {
                        val nodeId = stageIdsToIds[stage.id]!!
                        stageNodesMutable[stageIdsToIds[parentId]!!].add(nodeId)
                    }
                }
            }
        }

        stageOrdering = GraphUtils.findTopologicalOrdering(stageNodesMutable)
        nonCompleteCount = i
        stagesGraph = stageNodesMutable.map { it.toList() }.toTypedArray()

        i = 0
        fights.sortedBy { it.round }.sortedBy { stageOrdering[stageIdsToIds[it.stageId]!!] }.forEach { f ->
            fightIdToCategoryId[f.id] = f.categoryId
            fightIdToStageId[f.id] = f.stageId
            fightIdToDuration[f.id] = f.duration
            if (!fightIdsToIds.containsKey(f.id)) {
                fightIdsToIds[f.id] = i++
            }
        }

        val fightsGraphMutable = Array<MutableSet<Int>>(i) { HashSet() }


        fights.forEach { fight ->
            if (!fight.winFight.isNullOrBlank()) {
                fightsGraphMutable[fightIdsToIds[fight.id]!!].add(fightIdsToIds[fight.winFight]!!)
            }
            if (!fight.loseFight.isNullOrBlank()) {
                fightsGraphMutable[fightIdsToIds[fight.id]!!].add(fightIdsToIds[fight.loseFight]!!)
            }
            stageIdToFightIds[stageIdsToIds[fight.stageId]!!].add(fight.id)
        }
        completableFights = HashSet()
        fightsInDegree = GraphUtils.getIndegree(fightsGraphMutable)
        fightsInDegree.forEachIndexed { fight, degree ->
            if (degree == 0) {
                completableFights.add(fight)
            }
        }
        fightsOrdering = GraphUtils.findTopologicalOrdering(fightsGraphMutable)
        fightsGraph = fightsGraphMutable.map { it.toList() }.toTypedArray()
        completedFights = BooleanArray(i) { false }
        nonCompleteCount = i
    }

    fun completeFight(id: String) {
        val i = fightIdsToIds[id]
        if (i != null && !completedFights[i]) {
            completedFights[i] = true
            fightsGraph[i].forEach { adj ->
                if (!completedFights[adj]) {
                    if (--fightsInDegree[adj] <= 0) {
                        completableFights.add(adj)
                    }
                }
            }
            nonCompleteCount--
            completableFights.remove(i)
        }
    }

    fun getFightInDegree(id: String): Int {
        return fightsInDegree[fightIdsToIds[id]!!]
    }

    fun flushCompletableFights(fromFights: Set<String>): List<String> {
        return fromFights.filter { completableFights.contains(fightIdsToIds[it]) && !completedFights[fightIdsToIds.getValue(it)] }
                .sortedBy { fightIdsToIds[it]?.let { f -> fightsOrdering[f] } ?: Int.MAX_VALUE }
    }

    fun flushNonCompletedFights(fromFights: Set<String>): List<String> {
        return fromFights.filter { !completedFights[fightIdsToIds.getValue(it)] }
                .sortedBy { fightIdsToIds[it]?.let { f -> fightsOrdering[f] } ?: Int.MAX_VALUE }
    }

    fun getCategoryId(id: String): String {
        return fightIdToCategoryId[id]!!
    }
    fun getStageId(id: String): String {
        return fightIdToStageId[id]!!
    }
    fun getDuration(id: String): BigDecimal {
        return fightIdToDuration[id]!!
    }

    fun getStageFights(stageId: String): Set<String> {
        return if (stageIdsToIds.containsKey(stageId)) {
            stageIdToFightIds[stageIdsToIds[stageId]!!]
        } else {
            emptySet()
        }
    }

    fun getCategoryIdsToFightIds(): Map<String, Set<String>> {
        return categoryIdToFightIds
    }

    fun getNonCompleteCount(): Int {
        return nonCompleteCount
    }

}

File: saga-processor/src/main/scala/compman/compsrv/logic/service/schedule/StageGraph.scala
package compman.compsrv.logic.service.schedule

import com.google.common.collect.{BiMap, HashBiMap}
import compman.compsrv.logic.service.generate.CanFail
import compman.compsrv.model.dto.brackets.StageDescriptorDTO
import compman.compsrv.model.dto.competition.FightDescriptionDTO
import compman.compsrv.model.extension._

import scala.collection.mutable
import scala.collection.mutable.ArrayBuffer
import scala.language.reflectiveCalls

case class StageGraph private[schedule] (
                       private val fightsMap: Map[String, FightDescriptionDTO],
                       private val stagesGraph: Array[List[Int]],
                       private val fightsGraph: Array[List[Int]],
                       private val stageIdsToIds: BiMap[String, Int] = HashBiMap.create(),
                       private val fightIdsToIds: BiMap[String, Int] = HashBiMap.create(),
                       //  private val stageIdToFightIds: Array[mutable.Set[String]],
                       private val stageOrdering: Array[Int],
                       private val fightsOrdering: Array[Int],
                       private val nonCompleteCount: Int,
                       private val fightsInDegree: Array[Int],
                       private val completedFights: Set[Int],
                       private val completableFights: Set[Int],
                       private val categoryIdToFightIds: Map[String, List[String]],
                       private val stageIdsToFightIds: Array[List[String]]
                     ) {

  def getFightInDegree(id: String): Int = {
    fightsInDegree(fightIdsToIds.get(id))
  }

  def flushCompletableFights(fromFights: Set[String]): List[String] = {
    fromFights.filter { it =>
      completableFights.contains(fightIdsToIds.get(it)) && !completedFights(fightIdsToIds.get(it))
    }.toList.sortedFights
  }

  private implicit class ListStringOps(l: List[String]) {
    def sortedFights: List[String] = l.sortBy { it =>
      Option(fightIdsToIds.get(it)).map { f => fightsOrdering(f) }.getOrElse(Int.MaxValue)
    }
  }

  def flushNonCompletedFights(fromFights: Set[String]): List[String] = {
    fromFights.filter { it => completedFights(fightIdsToIds.get(it)) }.toList.sortedFights
  }

  def getCategoryId(id: String): String = {
    fightsMap(id).getCategoryId
  }

  def getStageId(id: String): String = {
    fightsMap(id).getStageId
  }

  def getDuration(id: String): BigDecimal = {
    fightsMap(id).getDuration
  }

  def getStageFights(stageId: String): List[String] = {
    if (stageIdsToIds.containsKey(stageId)) {
      stageIdsToFightIds(stageIdsToIds.get(stageId))
    }
    else {
      List.empty[String]
    }
  }

  def getCategoryIdsToFightIds: Map[String, List[String]] = {
    categoryIdToFightIds
  }

  def getNonCompleteCount: Int = {
    nonCompleteCount
  }
}

object StageGraph {

  import scala.jdk.CollectionConverters._

  def completeFight(id: String, stageGraph: StageGraph): StageGraph = {
    val i = stageGraph.fightIdsToIds.get(id)
    if (!stageGraph.completedFights(i)) {
      val newVertices = ArrayBuffer.empty[Int]
      stageGraph.fightsGraph(i).foreach { adj =>
        if (!stageGraph.completedFights(adj)) {
          stageGraph.fightsInDegree(adj) -= 1
          if (stageGraph.fightsInDegree(adj) <= 0) {
            newVertices.append(adj)
          }
        }
      }
      stageGraph.copy(completedFights = stageGraph.completedFights + i, nonCompleteCount = stageGraph.nonCompleteCount - 1, completableFights = (stageGraph.completableFights ++ newVertices) - i)
    }
  }

  private def createStagesGraph(
                                 stages: List[StageDescriptorDTO],
                                 stageIdsToIds: BiMap[String, Int]
                               ): Array[List[Int]] = {
    val stageNodesMutable = Array.fill(stageIdsToIds.keySet().size()) {
      mutable.HashSet.empty[Int]
    }
    stages.foreach { stage =>
      stage.getInputDescriptor.getSelectors.foreach { s =>
        val parentId = s.getApplyToStageId
        if (stageIdsToIds.containsKey(parentId)) {
          val nodeId = stageIdsToIds.get(stage.getId)
          stageNodesMutable(stageIdsToIds.get(parentId)).add(nodeId)
        }
      }
    }
    stageNodesMutable.map {
      _.toList
    }
  }

  private def createEntityIdsToIntIds[T <: {def getId: String}](stages: List[T]) = {
    val indices = stages.distinctBy(_.getId).zipWithIndex.map(p => (p._1.getId, p._2)).toMap
    HashBiMap.create(indices.asJava)
  }

  private def getFightsGraphAndStageIdToFightId(
                                                 fights: List[FightDescriptionDTO],
                                                 stagesNumber: Int,
                                                 fightsNumber: Int,
                                                 fightIdsToIds: BiMap[String, Int],
                                                 stageIdsToIds: BiMap[String, Int]
                                               ) = {
    val fightsGraphMutable = Array.fill(fightsNumber) {
      mutable.HashSet.empty[Int]
    }
    val stageIdToFightIds = Array.fill(stagesNumber) {
      mutable.HashSet.empty[String]
    }

    fights.foreach { fight =>
      if (fight.getWinFight != null) {
        fightsGraphMutable(fightIdsToIds.get(fight.getId)).add(fightIdsToIds.get(fight.getWinFight))
      }
      if (fight.getLoseFight != null) {
        fightsGraphMutable(fightIdsToIds.get(fight.getId)).add(fightIdsToIds.get(fight.getLoseFight))
      }
      stageIdToFightIds(stageIdsToIds.get(fight.getStageId)).add(fight.getId)
    }
    (fightsGraphMutable.map(_.toList), stageIdToFightIds.map(_.toList))
  }

  private def getCompletableFights(fightsInDegree: Array[Int]) = {
    val completableFights = mutable.HashSet.empty[Int]

    fightsInDegree.zipWithIndex.foreach { e =>
      val (degree, fight) = e
      if (degree == 0) {
        completableFights.add(fight)
      }
    }
    completableFights.toSet
  }

  def create(stages: List[StageDescriptorDTO], fights: List[FightDescriptionDTO]): CanFail[StageGraph] = {
    // we resolve transitive connections here (a -> b -> c  ~>  (a -> b, a -> c, b -> c))
    val stageIdsToIds = createEntityIdsToIntIds(stages)
    val stagesGraph = createStagesGraph(stages, stageIdsToIds)

    for {
      stageOrdering <- GraphUtils.findTopologicalOrdering(stagesGraph)
      sortedFights = fights.sortBy {
        _.roundOrZero
      }.sortBy { it => stageOrdering(stageIdsToIds.get(it.getStageId)) }
      fightsMap = sortedFights.groupMapReduce(_.getId)(identity)((a, _) => a)
      fightIdsToIds = createEntityIdsToIntIds(sortedFights)
      (fightsGraph, stageIdsToFightIds) =
        getFightsGraphAndStageIdToFightId(fights, stages.size, fightsMap.size, fightIdsToIds, stageIdsToIds)
      fightsInDegree = GraphUtils.getIndegree(fightsGraph.map(_.toSet))
      fightsOrdering <- GraphUtils.findTopologicalOrdering(fightsGraph)
      completableFights = getCompletableFights(fightsInDegree)
      completedFights = Set.empty[Int]
      categoryIdToFightIds = fights.groupMap(_.getCategoryId)(_.getId)
    } yield new StageGraph(
      stagesGraph = stagesGraph,
      fightsGraph = fightsGraph,
      stageIdsToIds = stageIdsToIds,
      fightIdsToIds = fightIdsToIds,
      fightsMap = fightsMap,
      stageOrdering = stageOrdering,
      fightsOrdering = fightsOrdering,
      nonCompleteCount = fightsMap.size,
      fightsInDegree = fightsInDegree,
      completedFights = completedFights,
      completableFights = completableFights,
      categoryIdToFightIds = categoryIdToFightIds,
      stageIdsToFightIds = stageIdsToFightIds
    )
  }

}
